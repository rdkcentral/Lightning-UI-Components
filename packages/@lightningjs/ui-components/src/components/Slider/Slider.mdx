import { Meta, Tile } from '@storybook/blocks';
import * as SliderStories from './Slider.stories';

<Meta of={SliderStories} />

<Title />

Progress control component, similar to the DOM's [input[type="range"]](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range)

## Source

https://github.com/rdkcentral/Lightning-UI-Components/blob/develop/packages/@lightningjs/ui-components/src/components/Slider/Slider.js

## Usage

Allows a user to select from a range of values using the `<-` and `->` keys.

```js
import { Slider, SliderLarge } from '@lightningjs/lightning-ui';

class Example extends lng.Component {
  static _template() {
    return {
      Slider: {
        type: Slider
      }
    };
  }
}
```

The `Slider` can be initialized with a `min`, `max` and current `value`. The `min` and `max` properties determine the bounds of the `Slider`. The `value` prop should be greater than or equal to `min` and less than or equal to `max`.

```js
import { Slider } from '@lightningjs/lightning-ui';

class Example extends lng.Component {
  static _template() {
    return {
      Slider: {
        type: Slider,
        min: 1,
        max: 5,
        value: 3
      }
    };
  }
}
```

The `step` property can be used to indicate _how much_ the slider should move by. In the below example, pressing `->` will jump by 10.

```js
class Example extends lng.Component {
  static _template() {
    return {
      Slider: {
        type: Slider,
        step: 10
      }
    };
  }
}
```

### Signal handling

React to changes to the `Slider` value using the `onChange` signal.

## API

### Properties

| name  | type   | readonly | default    | description            |
| ----- | ------ | -------- | ---------- | ---------------------- |
| max   | number | false    | 100        | upper bound of `value` |
| min   | number | false    | 0          | lower bound of `value` |
| step  | number | false    | 1          | +/- value on change    |
| value | number | false    | 0 or `min` | current value          |

### Style Properties

| name                 | type    | description                                                                                      |
| -------------------- | ------- | ------------------------------------------------------------------------------------------------ |
| arrowAlphaValue      | number  | alpha value of the arrows if value is not equal to the min/max value                             |
| arrowAlphaValueLimit | number  | alpha value of the arrows when value is equal to the min/max value and slider is not disabled    |
| arrowColor           | string  | color of the arrows                                                                              |
| arrowHeight          | number  | height of the arrows                                                                             |
| arrowSpacing         | number  | space between the arrow and the slider                                                           |
| arrowWidth           | number  | width of the arrows                                                                              |
| circleColor          | string  | color of the knob                                                                                |
| containerHeight      | number  | height of the container that holds the slider                                                    |
| minWidth             | number  | minimum width of the slider container                                                            |
| progressStyle        | object  | style inherited from progressBar [See ProgressBar documentation](../ProgressBar/ProgressBar.mdx) |
| showArrows           | boolean | if true, will display the left and right arrows on the slider                                    |
| showKnob             | boolean | if true, will display the focused circle on the current value of the slider                      |
| sliderHeight         | number  | height of the slider                                                                             |

### Signals

#### onChange(value:number, signal:Signal)

Fires on key Left/Right events when `Slider` is in focus.
