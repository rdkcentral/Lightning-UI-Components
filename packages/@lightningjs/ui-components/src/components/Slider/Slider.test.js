/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { makeCreateComponent, fastForward } from '@lightningjs/ui-components-test-utils';
import { degreesToRadians } from '../../utils/index.js';
import Slider from './index.js';
import { jest } from '@jest/globals';

const createSlider = makeCreateComponent(Slider);

describe('Slider', () => {
  let slider, testRenderer;

  beforeEach(() => {
    [slider, testRenderer] = createSlider();
    testRenderer.update();
  });

  afterEach(() => {
    slider = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to slider value', () => {
    slider.value = 50;
    testRenderer.forceAllUpdates();
    expect(slider.announce).toBe('50');
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    slider.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(slider.announce).toBe(overrideString);
  });

  it('reannounces on progress change', () => {
    slider.fireAncestors = jest.fn();
    slider.value = 75;
    testRenderer.forceAllUpdates();
    expect(slider.fireAncestors).toHaveBeenCalledWith('$announce', '75');
  });

  it('sets properties', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 0.5,
      value: 5
    });
    testRenderer.update();
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the initial LeftBar and Circle textures without smoothing', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 5
    });

    expect(slider._SliderBar.w).toBeGreaterThan(0);
  });

  it('Circle x position when value is greater than the max value', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 15
    });

    expect(slider._Circle.x).toEqual(slider._calculatedSliderWidth);
  });

  it('Circle x position when value is less than the min value', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 0
    });

    expect(slider._Circle.x).toEqual((1 / 10) * slider._calculatedSliderWidth);
  });

  it('Circle x position and progress bar width when minimum value is negative', () => {
    [slider, testRenderer] = createSlider({
      min: -10,
      max: 10,
      step: 1,
      value: 3
    });

    expect(slider._Circle.x).toEqual(
      ((slider.value - slider.min) / (slider.max - slider.min)) *
        slider._calculatedSliderWidth
    );
    expect(slider._SliderBar.progress).toEqual(
      (slider.value - slider.min) / (slider.max - slider.min)
    );
  });

  it('Circle is only alpha-ed on if Slider is focused and showKnob is true', () => {
    testRenderer.unfocus();
    testRenderer.forceAllUpdates();
    expect(slider._Circle.alpha).toEqual(0);

    [slider, testRenderer] = createSlider({
      style: { showKnob: true }
    });
    testRenderer.focus();
    testRenderer.forceAllUpdates();
    expect(slider.style.showKnob).toBe(true);
    expect(slider._Circle.alpha).toEqual(1);

    [slider, testRenderer] = createSlider({
      style: { showKnob: false }
    });
    testRenderer.forceAllUpdates();
    expect(slider.style.showKnob).toBe(false);
    expect(slider._Circle.alpha).toEqual(0);
  });

  it('Alpha value of arrow when value is equal to max value', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 1
    });
    testRenderer.forceAllUpdates();
    expect(slider._LeftArrow.alpha).toEqual(0.5);
  });

  it('Alpha value of arrow when mode is set to disabled', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 1,
      mode: 'disabled'
    });
    testRenderer.forceAllUpdates();
    expect(slider._LeftArrow.alpha).toEqual(0);
  });

  it('Width larger than knob and arrows', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 2,
      w: 300
    });
    expect(slider._calculatedSliderWidth).toEqual(
      300 - slider.style.arrowSpacing * 2 - slider.style.arrowWidth * 2
    );
  });

  it('handle left override function', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 2,
      disabled: false,
      onLeft: () => {
        return false;
      }
    });
    slider._handleLeft();
    expect(slider.onLeft()).toEqual(false);
  });

  it('handle left when mode is set to disabled', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 2,
      mode: 'disabled'
    });
    expect(slider._handleLeft()).toEqual(false);
  });

  it('handle right override function', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 2,
      disabled: false,
      onRight: () => {
        return false;
      }
    });
    slider._handleRight();
    expect(slider.onRight()).toEqual(false);
  });

  it('handle right when mode is set to disabled', () => {
    [slider, testRenderer] = createSlider({
      min: 1,
      max: 10,
      step: 1,
      value: 2,
      mode: 'disabled'
    });
    testRenderer.forceAllUpdates();
    expect(slider._handleRight()).toEqual(false);
  });

  describe('key handling', () => {
    describe('right', () => {
      beforeEach(() => {
        [slider, testRenderer] = createSlider({
          max: 2,
          value: 0
        });
      });

      it('increases the value', () => {
        slider._handleRight();
        expect(slider.value).toEqual(1);
      });

      it('stops increasing at max value', () => {
        slider._handleRight();
        slider._handleRight();
        expect(slider.value).toEqual(2);
        slider._handleRight();
        expect(slider.value).toEqual(2);
      });

      it('updates the Bar and Circle textures with smoothing', async () => {
        [slider, testRenderer] = createSlider(
          { max: 2, value: 1 },
          { spyOnMethods: ['_update'] }
        );
        await slider.__updateSpyPromise;
        expect(slider._SliderBar._Progress.w).toEqual(0);

        testRenderer.keyPress('Right');
        await slider.__updateSpyPromise;

        fastForward(slider._SliderBar);
        testRenderer.update();
        const updatedWidth = testRenderer.getInstance()._SliderBar._Progress.w;
        expect(updatedWidth).toBeGreaterThan(0);
      });

      it('should call onRight function if passed as a property to Slider', () => {
        const onRightSpy = jest.fn();
        slider.onRight = onRightSpy;
        expect(onRightSpy).not.toHaveBeenCalled();

        testRenderer.keyPress('Right');

        expect(onRightSpy).toHaveBeenCalled();
      });

      it('should ignore the key event if the slider is vertical', () => {
        const onRightSpy = jest.fn();
        slider.onRight = onRightSpy;
        slider.vertical = true;

        expect(onRightSpy).not.toHaveBeenCalled();

        testRenderer.keyPress('Right');

        expect(onRightSpy).not.toHaveBeenCalled();
      });
    });

    describe('up', () => {
      beforeEach(() => {
        [slider, testRenderer] = createSlider({
          vertical: true,
          max: 2,
          value: 2
        });
      });

      it('decreases the value', () => {
        testRenderer.keyPress('Up');
        expect(slider.value).toEqual(1);
      });

      it('stops decreasing at the min value', () => {
        testRenderer.keyPress('Up');
        testRenderer.keyPress('Up');
        expect(slider.value).toEqual(0);
        testRenderer.keyPress('Up');
        expect(slider.value).toEqual(0);
      });

      it('updates the Bar and Circle textures with smoothing', async () => {
        [slider, testRenderer] = createSlider(
          {
            max: 2,
            value: 2
          },
          {
            spyOnMethods: ['_update']
          }
        );
        await slider.__updateSpyPromise;
        const initialWidth = slider._SliderBar.w;
        expect(slider._SliderBar.w).toEqual(initialWidth);

        testRenderer.keyPress('Left');
        await slider.__updateSpyPromise;

        fastForward(slider._SliderBar);
        testRenderer.update();
        const updatedWidth = slider._SliderBar._Progress.w;
        expect(updatedWidth).toBeLessThan(initialWidth);
      });

      it('should call onUp function if passed as a property to Slider', () => {
        const onUpSpy = jest.fn();
        slider.onUp = onUpSpy;
        expect(onUpSpy).not.toHaveBeenCalled();

        testRenderer.keyPress('Up');

        expect(onUpSpy).toHaveBeenCalled();
      });

      it('should ignore the key event if the slider is not vertical', () => {
        const onUpSpy = jest.fn();
        slider.onUp = onUpSpy;
        slider.vertical = false;

        expect(onUpSpy).not.toHaveBeenCalled();

        testRenderer.keyPress('Up');

        expect(onUpSpy).not.toHaveBeenCalled();
      });
    });

    describe('down', () => {
      beforeEach(() => {
        [slider, testRenderer] = createSlider({
          vertical: true,
          max: 2,
          value: 0
        });
      });

      it('increases the value', () => {
        testRenderer.keyPress('Down');
        expect(slider.value).toEqual(1);
      });

      it('stops increasing at max value', () => {
        testRenderer.keyPress('Down');
        testRenderer.keyPress('Down');
        expect(slider.value).toEqual(2);
        testRenderer.keyPress('Down');
        expect(slider.value).toEqual(2);
      });

      it('should call onDown function if passed as a property to Slider', () => {
        const onDownSpy = jest.fn();
        slider.onDown = onDownSpy;
        expect(onDownSpy).not.toHaveBeenCalled();

        testRenderer.keyPress('Down');

        expect(onDownSpy).toHaveBeenCalled();
      });

      it('should ignore the key event if the slider is vertical', () => {
        const onDownSpy = jest.fn();
        slider.onDown = onDownSpy;
        slider.vertical = false;

        expect(onDownSpy).not.toHaveBeenCalled();

        testRenderer.keyPress('Down');

        expect(onDownSpy).not.toHaveBeenCalled();
      });
    });
  });

  describe('onChange signal', () => {
    beforeEach(() => {
      Slider.prototype.signal = jest.fn();
      [slider, testRenderer] = createSlider({
        value: 50
      });
    });

    it('is fired on init', () => {
      expect(slider.signal).toBeCalledWith('onChange', 50, slider);
    });

    it('is fired on key left', async () => {
      [slider, testRenderer] = createSlider(
        { value: 50 },
        { spyOnMethods: ['_update'] }
      );
      await slider.__updateSpyPromise;
      slider._handleLeft();
      await slider.__updateSpyPromise;
      expect(slider.signal).toBeCalledWith('onChange', 49, slider);
    });

    it('is fired on key right', async () => {
      [slider, testRenderer] = createSlider(
        { value: 50 },
        { spyOnMethods: ['_update'] }
      );
      await slider.__updateSpyPromise;
      slider._handleRight();
      await slider.__updateSpyPromise;
      expect(slider.signal).toBeCalledWith('onChange', 51, slider);
    });
  });

  describe('when a vertical slider is rendered', () => {
    beforeEach(() => {
      [slider, testRenderer] = createSlider({ vertical: true });
    });

    it('should rotate the default slider by 90 degrees', () => {
      expect(slider.rotation).toBe(degreesToRadians(90));
    });
  });
});
