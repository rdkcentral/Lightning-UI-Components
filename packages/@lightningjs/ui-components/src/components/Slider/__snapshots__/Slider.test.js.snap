// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders 1`] = `
{
  "Component": {
    "Container": {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Bar": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Circle": {
              "active": false,
              "alpha": 0,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": false,
              "flex": false,
              "flexItem": false,
              "h": 20,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0.5,
              "mountX": 0.5,
              "mountY": 0.5,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Circle",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": [
                "Circle",
              ],
              "type": "Knob",
              "visible": true,
              "w": 20,
              "x": 0,
              "y": 13,
              "zIndex": 5,
            },
            "SliderBar": {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "children": {
                "Bar": {
                  "active": true,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Bar",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Bar",
                  ],
                  "texture": {
                    "type": "StaticCanvasTexture",
                  },
                  "visible": true,
                  "w": 0,
                  "x": 0,
                  "y": 0,
                  "zIndex": 1,
                },
                "Progress": {
                  "active": false,
                  "alpha": 0,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": false,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Progress",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Progress",
                  ],
                  "texture": {
                    "type": "StaticCanvasTexture",
                  },
                  "visible": true,
                  "w": 0,
                  "x": 0,
                  "y": 0,
                  "zIndex": 2,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 10,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0.5,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "SliderBar",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": [
                "SliderBar",
              ],
              "type": "ProgressBar",
              "visible": true,
              "w": 208,
              "x": 0,
              "y": 12,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Bar",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "Bar",
          ],
          "visible": true,
          "w": 0,
          "x": 50,
          "y": 0,
          "zIndex": 0,
        },
        "LeftArrow": {
          "active": true,
          "alpha": 0.5,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": "fff8f7fa",
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 40,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "LeftArrow",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "LeftArrow",
          ],
          "texture": {
            "h": 40,
            "src": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAGKADAAQAAAABAAAAGAAAAADiNXWtAAAAi0lEQVRIDWNgGAWjIfD//38JID5Fk5AAGqwKxPeA+D/VLQCaaQLEr0CGgwBVLQCa5wbEn0EGwwDVLAAaGA3Ev2AGw2iqWAA0rBiI/8EMRaYptgBoWDeygehsci1gIlcjWfqArqZdEMFcBLSEdpGMZAntkimSJbTLaEiW0K6oQLKEdoUdzJJRemiHAAD4n+yzPWCs7QAAAABJRU5ErkJggg==",
            "type": "CustomImageTexture",
            "w": 40,
          },
          "type": "Icon",
          "visible": true,
          "w": 40,
          "x": 0,
          "y": [Function],
          "zIndex": 0,
        },
        "RightArrow": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": "fff8f7fa",
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 40,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "RightArrow",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "RightArrow",
          ],
          "texture": {
            "h": 40,
            "src": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAGKADAAQAAAABAAAAGAAAAADiNXWtAAAAg0lEQVRIDWNgGAWjIYArBP7//38KiCVwyVMsDjQcBO4BsSrFhmEzAGw8hHgFpEywqaFIDMkCEPMzELtRZCC6ZjQLQNxfQByNro5sPhYLQEL/gLiYbEORNeKwACbcDVPLBGMMOhrmVDSapkFE00imaTKlaUajaVFB28Ju0CXrUQfhDAEAEgHss6NhpLQAAAAASUVORK5CYII=",
            "type": "CustomImageTexture",
            "w": 40,
          },
          "type": "Icon",
          "visible": true,
          "w": 40,
          "x": 268,
          "y": [Function],
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 24,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Container",
      ],
      "visible": true,
      "w": 20,
      "x": 0,
      "y": [Function],
      "zIndex": 0,
    },
  },
}
`;

exports[`Slider sets properties 1`] = `
{
  "Component": {
    "Container": {
      "active": true,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Bar": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Circle": {
              "active": false,
              "alpha": 0,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": false,
              "flex": false,
              "flexItem": false,
              "h": 20,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0.5,
              "mountX": 0.5,
              "mountY": 0.5,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Circle",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": [
                "Circle",
              ],
              "type": "Knob",
              "visible": true,
              "w": 20,
              "x": 104,
              "y": 13,
              "zIndex": 5,
            },
            "SliderBar": {
              "active": true,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "children": {
                "Bar": {
                  "active": true,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Bar",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Bar",
                  ],
                  "texture": {
                    "type": "StaticCanvasTexture",
                  },
                  "visible": true,
                  "w": 0,
                  "x": 0,
                  "y": 0,
                  "zIndex": 1,
                },
                "Progress": {
                  "active": true,
                  "alpha": 1,
                  "attached": true,
                  "boundsMargin": null,
                  "clipping": false,
                  "color": 4294967295,
                  "enabled": true,
                  "flex": false,
                  "flexItem": false,
                  "h": 0,
                  "isComponent": undefined,
                  "mount": 0,
                  "mountX": 0,
                  "mountY": 0,
                  "pivot": 0.5,
                  "pivotX": 0.5,
                  "pivotY": 0.5,
                  "ref": "Progress",
                  "renderOfScreen": undefined,
                  "renderToTexture": false,
                  "scale": 1,
                  "scaleX": 1,
                  "scaleY": 1,
                  "state": undefined,
                  "tag": [Function],
                  "tags": [
                    "Progress",
                  ],
                  "texture": {
                    "type": "StaticCanvasTexture",
                  },
                  "visible": true,
                  "w": 8.320000187412864,
                  "x": 0,
                  "y": 0,
                  "zIndex": 2,
                },
              },
              "clipping": false,
              "color": 4294967295,
              "enabled": true,
              "flex": false,
              "flexItem": false,
              "h": 10,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0.5,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "SliderBar",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": [
                "SliderBar",
              ],
              "type": "ProgressBar",
              "visible": true,
              "w": 208,
              "x": 0,
              "y": 12,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Bar",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "Bar",
          ],
          "visible": true,
          "w": 0,
          "x": 50,
          "y": 0,
          "zIndex": 0,
        },
        "LeftArrow": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": "fff8f7fa",
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 40,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "LeftArrow",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "LeftArrow",
          ],
          "texture": {
            "h": 40,
            "src": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAGKADAAQAAAABAAAAGAAAAADiNXWtAAAAi0lEQVRIDWNgGAWjIfD//38JID5Fk5AAGqwKxPeA+D/VLQCaaQLEr0CGgwBVLQCa5wbEn0EGwwDVLAAaGA3Ev2AGw2iqWAA0rBiI/8EMRaYptgBoWDeygehsci1gIlcjWfqArqZdEMFcBLSEdpGMZAntkimSJbTLaEiW0K6oQLKEdoUdzJJRemiHAAD4n+yzPWCs7QAAAABJRU5ErkJggg==",
            "type": "CustomImageTexture",
            "w": 40,
          },
          "type": "Icon",
          "visible": true,
          "w": 40,
          "x": 0,
          "y": [Function],
          "zIndex": 0,
        },
        "RightArrow": {
          "active": true,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": "fff8f7fa",
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 40,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "RightArrow",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "RightArrow",
          ],
          "texture": {
            "h": 40,
            "src": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAGKADAAQAAAABAAAAGAAAAADiNXWtAAAAg0lEQVRIDWNgGAWjIYArBP7//38KiCVwyVMsDjQcBO4BsSrFhmEzAGw8hHgFpEywqaFIDMkCEPMzELtRZCC6ZjQLQNxfQByNro5sPhYLQEL/gLiYbEORNeKwACbcDVPLBGMMOhrmVDSapkFE00imaTKlaUajaVFB28Ju0CXrUQfhDAEAEgHss6NhpLQAAAAASUVORK5CYII=",
            "type": "CustomImageTexture",
            "w": 40,
          },
          "type": "Icon",
          "visible": true,
          "w": 40,
          "x": 268,
          "y": [Function],
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 24,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Container",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Container",
      ],
      "visible": true,
      "w": 20,
      "x": 0,
      "y": [Function],
      "zIndex": 0,
    },
  },
}
`;
